jme3-terrain/src/main/java/com/jme3/terrain/geomipmap/TerrainLodControl.java

```diff
@@ -176,53 +228,110 @@ protected void controlUpdate(float tpf) {
      * It will clear up any threads it had.
      */
     public void detachAndCleanUpControl() {
-        if (executor != null)
-            executor.shutdownNow();
+
+        if (indexer != null) {
+            indexer.cancel(true);
+            indexer = null;
+        }
+
         getSpatial().removeControl(this);
     }
 
     // do all of the LOD calculations
-    protected void updateLOD(List<Vector3f> locations, LodCalculator lodCalculator) {
-        if(getSpatial() == null){
+    protected void updateLOD(final LodCalculator lodCalculator) {
+
+        if (getSpatial() == null || camera == null) {
             return;
         }
 
         // update any existing ones that need updating
         updateQuadLODs();
 
-        if (lodCalculator.isLodOff()) {
-            // we want to calculate the base lod at least once
-            if (lodOffCount == 1)
-                return;
-            else
-                lodOffCount++;
-        } else
-            lodOffCount = 0;
+        if (updateLodOffCount(lodCalculator)) {
+            return;
+        }
+
+        final Vector3f currentLocation = camera.getLocation();
 
-        if (lastCameraLocations != null) {
-            if (!forceUpdate && lastCameraLocationsTheSame(locations) && !lodCalculator.isLodOff())
-                return; // don't update if in same spot
-            else
-                lastCameraLocations = cloneVectorList(locations);
-            forceUpdate = false;
+        if (!forceUpdate && previousCameraLocation.equals(currentLocation) && !lodCalculator.isLodOff()) {
+            return; // don't update if in same spot
+        } else {
+            previousCameraLocation.set(currentLocation);
         }
-        else {
-            lastCameraLocations = cloneVectorList(locations);
+
+        forceUpdate = false;
+
+        if (!lodCalcRunning.compareAndSet(false, true)) {
+            return;
+        }
+
+        prepareTerrain();
+
+        final SafeArrayList<Vector3f> locations = new SafeArrayList<>(Vector3f.class, 1);
+        locations.add(currentLocation);
+
+        final TerrainExecutorService executorService = TerrainExecutorService.getInstance();
+        indexer = executorService.submit(createLodUpdateTask(locations, lodCalculator));

```

Comment:
Ho .... for some reason github didn't expand the content of the file so I missed it...